server/src/daemon/service.rs

```diff
@@ -0,0 +1,311 @@
+                }
+            }
+        }
+
+        if locked.is_empty() {
+            locked.push(OwnedObjectPath::default());
+        }
+
+        let prompt = Prompt::default(); // temporarily
+
+        Ok((locked, prompt))
+    }
+
+    pub async fn get_secrets(
+        &self,
+        paths: Vec<OwnedObjectPath>,
+        session: ObjectPath<'_>,
+    ) -> Result<HashMap<OwnedObjectPath, SecretInner>> {
+        let mut secrets = HashMap::with_capacity(paths.len());
+        for collection in self.collections.read().await.iter() {
+            let items = collection.items.read().await;
+            for item in items.iter() {
+                for path in paths.iter() {
+                    if item.path() == path.as_ref() {
+                        let secret = item.secret(session.clone()).await?;
+                        secrets.insert(path.to_owned(), secret);
+                    }
+                }
+            }
+        }
+        Ok(secrets)
+    }
+
+    pub async fn read_alias(&self, name: &str) -> ObjectPath {
+        self.collections
+            .read()
+            .await
+            .iter()
+            .find_map(|c| {
+                if c.label() == name {
+                    Some(c.path().to_owned())
+                } else {
+                    None
+                }
+            })
+            .unwrap_or_default()
+    }
+
+    pub async fn set_alias(
+        &self,
+        #[zbus(signal_context)] ctxt: SignalContext<'_>,
+        alias: &str,
+        path: ObjectPath<'_>,
+    ) -> Result<()> {
+        match self
+            .collections
+            .read()
+            .await
+            .iter()
+            .find(|c| c.path() == path)
+        {
+            Some(collection) => {
+                collection.set_alias(&ctxt, alias).await?;
+                Ok(())
+            }
+            None => {
+                tracing::error!("Collection {path} not found");
+                Err(ServiceError::NoSuchObject)
+            }
+        }
+    }
+
+    #[zbus(property, name = "Collections")]
+    pub async fn collections(&self) -> Vec<ObjectPath> {
+        self.collections
+            .read()
+            .await
+            .iter()
+            .map(|collection| collection.path().to_owned())
+            .collect()
+    }
+
+    #[zbus(signal)]
+    pub async fn collection_created(
+        ctxt: &SignalContext<'_>,
+        collection: ObjectPath<'_>,
+    ) -> zbus::Result<()>;
+
+    #[zbus(signal)]
+    pub async fn collection_deleted(
+        ctxt: &SignalContext<'_>,
+        collection: ObjectPath<'_>,
+    ) -> zbus::Result<()>;
+
+    #[zbus(signal)]
+    pub async fn collection_changed(
+        ctxt: &SignalContext<'_>,
+        collection: ObjectPath<'_>,
+    ) -> zbus::Result<()>;
+}
+
+impl Service {
+    pub async fn new() -> Self {
+        Self {
+            collections: RwLock::new(Vec::new()),
+            keyring: Arc::new(Keyring::load_default().await.unwrap()),
```

Comment:
The unwrap here makes me nervous, should fallback to create the default if not available.

Reply:
Fixed with https://github.com/bilelmoussaoui/oo7/pull/73/commits/b2abf3c42aa967882e3e34172c82ff06e3bbb608

Reply:
The todo! is not a fix.

Reply:
Even load_default call is actually wrong. You need to call `.load` instead as you don't use the portal's key to encrypt the keyring.

Reply:
Understood, thanks!

Reply:
> The todo! is not a fix.

Please see https://github.com/bilelmoussaoui/oo7/pull/73#discussion_r1521541858 :)

Reply:
I have started a new discussion [here](https://github.com/bilelmoussaoui/oo7/pull/73#discussion_r1523242017)